### ISSUE TITLE
Temporary Road Closures Datasets - Direct API Consumption

### ISSUE THREAD

**Issue by al-tabb** [bcgov/smk-tlink#204]
Hi @NicoledeGreef,

As you know the Temporary Road Closures datasets require by far the most upkeep in the TRP app.  With the goal of reducing this effort on your end (in the long term) I'm suggesting that the application directly consume the GeoJson data from TransLink's API.  This will allow us be more responsive with updating the information as these road closures are added, removed and altered.

These are the endpoints:
https://api.regionalroads.com/simplewfs/?service=WFS&version=1.0.0&request=GetFeature&typeName=GM_TemporaryRoadClosures&outputFormat=application/json

https://api.regionalroads.com/simplewfs/?service=WFS&version=1.0.0&request=GetFeature&typeName=GM_TemporaryRoadClosuresExtent&outputFormat=application/json

One more wrinkle related to these datasets:  I believe that previously our datetime data was stored in UTC format and was not adjusted to local time.  As a result of changes to the workflow on my end the datetimes are now converted to UTC from the local time.  So I believe if we don't make an adjustment they will be 7-8 hours off.  Is it possible to add logic into the TRP app to perform this conversion? (i.e. Convert from UTC to local time).  If needed, I'd be happy to share my code for how I'm performing this conversion in my apps.

I recognize that this will involve coding changes on your end so I'm interested to hear what you think about this idea and the anticipated level of effort.

Thanks,

Alan

Tagging @gk-tl for awareness.

**Comment by NicoledeGreef** [comment]
Hi @qqnluaq  - would you please review this item and provide an estimate?

**Comment by qqnluaq** [comment]
@NicoledeGreef This would apply to only those 2 layers? I think effort involved would be about 3 days

**Comment by al-tabb** [comment]
@NicoledeGreef @qqnluaq 
Thanks for providing the estimate.  As this may require an amendment to the funding agreement it might make sense to bundle a few other items into this work.  Are you able to provide estimates on the following?  Please also let me know if you have the capacity to take these on.

- Extend the above request to directly consume all cartographic layers from TransLink's API
- Update colours / symbology of Truck Routes layer
- Change the default basemap

Thanks,

Alan

**Comment by NicoledeGreef** [comment]
@qqnluaq  - how many days to implement based on new info from @al-tabb?

**Comment by qqnluaq** [comment]
- Extend the above request to directly consume all cartographic layers from TransLink's API
1 - 3 days. 

- Update colours / symbology of Truck Routes layer
2 hours, but depends on complexity of new styling

- Change the default basemap
5 minutes, if changing to one of the existing basemap choices.

**Comment by qqnluaq** [comment]
@al-tabb @NicoledeGreef 
I've started this work, and immediately ran into a problem.
A request like `https://api.regionalroads.com/simplewfs/?service=WFS&version=1.0.0&request=GetFeature&typeName=GM_Cardlock&outputFormat=application/json` will fail when it is made from the context of the browser, with this error:
![image](https://user-images.githubusercontent.com/20274603/152033941-b5ee736f-2538-4e03-8a60-fbe207b483a3.png)
This is due the server at `api.regionalroads.com` not being configured for CORS.

**Comment by al-tabb** [comment]
@qqnluaq I just checked and I'm currently only allowing my app in the access-control-allow-origin.  What is the url that will be requesting the resource?

**Comment by qqnluaq** [comment]
@al-tabb since this will be publicly accessible, the allowed origin has to be `*`

**Comment by al-tabb** [comment]
The only front end apps that will be making requests to the API will be my internal TransLink apps and the TRP.  I should just be able to allow those two origins and it will work.  I've added https://translink-d.apps.gov.bc.ca to the allowed hosts for my dev API.  
Can you give it a try?  The dev version of the URL you're trying is here:

https://devapi.regionalroads.com/simplewfs/?service=WFS&version=1.0.0&request=GetFeature&typeName=GM_Cardlock&outputFormat=application/json

**Comment by qqnluaq** [comment]
@al-tabb  I get a 403 trying to access `https://devapi.regionalroads.com/simplewfs/?service=WFS&version=1.0.0&request=GetFeature&typeName=GM_Cardlock&outputFormat=application/json`

**Comment by al-tabb** [comment]
Had to change the access settings on the server.. give it a try now

**Comment by qqnluaq** [comment]
Still get a CORS error, the access-contorl-allow-origin header isn't coming back with the response.
![image](https://user-images.githubusercontent.com/20274603/152051409-96cbd294-409f-45fc-b925-c7d3cb1872cd.png)

( i updated my `/etc/hosts` file to spoof `translink-d.apps.gov.bc.ca` as localhost so i can test my local build)

**Comment by al-tabb** [comment]
Perhaps its the port (7443)?  I tried adding the port to the url, does it work now?

**Comment by qqnluaq** [comment]
Yup, that did it!
![image](https://user-images.githubusercontent.com/20274603/152052623-f91b6df6-3711-4417-88ae-7c3888c05df6.png)

**Comment by al-tabb** [comment]
OK awesome

**Comment by qqnluaq** [comment]
@al-tabb Ive identified an issue with the regionalroads api: 
I wasn't seeing anything on the map for this dataset `https://devapi.regionalroads.com/simplewfs/?service=WFS&version=1.0.0&request=GetFeature&outputFormat=application/json&typeName=GM_IndustrialAreas`.
The problem is that the features are being returned in the wrong coordinate system. The CRS attached to the feature collection says `EPSG::4326` but the coordinates are not geographic, they look like BC Albers if I had to guess.
The CRS attached to the feature set should reflect the actual projection being used if I am to reproject it to lat-long (which is what SMK expects).
I haven't checked all the datasets yet, but so far GM_Cardlock is returning data in the correct projection.
GM_DangerousGoodsRestrictions, GM_DowntownVancouver, GM_IndustrialAreas do not.
I tried adding srsName=EPSG:4326 to the request, but it made no difference.

**Comment by al-tabb** [comment]
Looks like I hadn't set up the reproject option in the geoserver service correctly.  They should all be in 4326 now.  Let me know if you have any more issues.

**Comment by qqnluaq** [comment]
I'm getting access denied for GM_OverheadStructures

**Comment by al-tabb** [comment]
Right, I forgot to make that public.  You should be good to go now.

**Comment by qqnluaq** [comment]
@NicoledeGreef  OK, I've committed a new version on branch develop-bjubb. This version will load all layers from the translink api. As well, this commit partially address #210: there is now a file  `config/site/config.js` that contains site-specific configuration for the app. It contains one key: `regionalRoadsApiUrl` which is the base Url for the translink API.

**Comment by NicoledeGreef** [comment]
thanks @qqnluaq!  we're starting a new sprint tomorrow and we'll continue with #210

**Comment by al-tabb** [comment]
@NicoledeGreef Perhaps its a good idea to have a meeting with yourself, @qqnluaq and @gk-tl to discuss the symbology changes and basemap change.  We are not in a huge rush to have these changes completed so I'm thinking of targeting a meeting in early March.  Does this work for you?

**Comment by dgboss** [comment]
@al-tabb I am working on deploying the change introduced by @qqnluaq and am running into a CORS issue. Could you please check if the translink-d.apps.gov.bc.ca is an allowed origin (without the 7443 port)?

Similarly, translink-t.apps.gov.bc.ca will need to be an allowed origin for stgapi.regionalroads.com and translink.apps.gov.bc.ca an allowed origin for api.regionalroads.com.

Thank you!

![image](https://user-images.githubusercontent.com/7094320/156829070-3c687909-63ae-4e08-845f-ca60800326e5.png)

**Comment by al-tabb** [comment]
@dgboss You should be good to go now.  Let me know if you have any more issues.
@qqnluaq Is it required that the dev environment allows the origin on port 7443 on an ongoing basis?

**Comment by dgboss** [comment]
Thank you @al-tabb, it is working well now. @qqnluaq's changes are now published to https://translink-d.apps.gov.bc.ca/trp/. If you visit the page, open your developer tools and view the network tab, you won't see GeoJSON being downloaded anymore, you will data coming in via WFS requests.

**Comment by al-tabb** [comment]
Looks good, thanks @dgboss I just tested it out by making a change to the temporary road closures datasets in my dev environment.  The changes were immediately reflected on page reload.

**Comment by dgboss** [comment]
@al-tabb that's great news! May I promote this to our Test environment (which will use your WFS services at stgapi)?

@qqnluaq Do you anticipate any further work being required on your develop-jubb branch? Or is it ready to merge into our main branch?

**Comment by qqnluaq** [comment]
I don't anticipate further work. Go ahead and merge. Please note my comment in https://github.com/bcgov/smk-tlink/issues/204#issuecomment-1029537829 regarding `config.js`

**Comment by al-tabb** [comment]
Sure, please promote to test whenever you're ready.

**Comment by dgboss** [comment]
@qqnluaq Thanks for the heads-up. I have modified our deployment process to copy the correct version of config.js into each of the environments. The correlation between environments looks like this:

https://translink-d.apps.gov.bc.ca/trp/ uses devapi.regionalroads.com
https://translink-t.apps.gov.bc.ca/trp/ uses stgapi.regionalroads.com
https://translink.apps.gov.bc.ca/trp/ will use api.regionalroads.com

**Comment by dgboss** [comment]
@al-tabb I have deployed the new code to our Test environment. 

https://translink-t.apps.gov.bc.ca/trp/ is now consuming your WFS services instead of downloading geojson.

Please let me know when we are good to go with a deployment to the production environment. Also, please let me know if there is a specific time window you would prefer the production update to occur in (for example, early morning, late afternoon, early evening or any specific day, etc). Thanks!

**Comment by al-tabb** [comment]
Looks good, feel free to deploy to production whenever works best for you.  Thanks.

**Comment by dgboss** [comment]
Thanks @al-tabb. I will target tomorrow morning (Mar 8, between 7:30am - 8:00am) for the update. In theory there should be no discernible downtime. I'll add a note here once the update is complete.

**Comment by dgboss** [comment]
@al-tabb the update in production is complete.

**Comment by al-tabb** [comment]
Thanks @dgboss.  Looks like everything is working well.
